### PR TITLE
MMI Instruction Implementation

### PR DESCRIPTION
**Pull request by astrelsky**
I have added in the instructions in my repo to this one. The cspec still is not quite right. "BADSPACEBASE" still appears in some functions when the sp is being stored/retrieved. It comes and goes when changing the stackpointer register to/from sp_lo/sp.

### CODE REVIEW

**Comment by beardypig** [comment]
I'll need some time to look at this properly. My only concern is that the special cases of `PADDUB`, etc. were done to improve decompilation - if they operate on the 128 bit registers you loose that improvement and get poor decompilation.

**Comment by astrelsky** [comment]
The special cases are still there. The varnode size limitation doesn't affect a direct = so I changed the special cases of moveq to be RD128 = RS128src;

Take all the time you need, it's a lot.

**Comment by beardypig** [comment]
It's not that there is a limit to the size of a varnode, it's that there is a limit to the size of a constant varnode. Which does cause some ugly decompilation, when I have a moment I'll check the specifics.

**Comment by astrelsky** [comment]
> It's not that there is a limit to the size of a varnode, it's that there is a limit to the size of a constant varnode. Which does cause some ugly decompilation, when I have a moment I'll check the specifics.

I see. Will definitely have to add as many special cases as possible. It doesn't come out too bad. If you run a search for memchr you will quickly come across some mmi instructions. In all honesty, considering what the instructions actually do, it's the unnecessary typecasting that is the most ugly.

File: data/languages/r5900.cspec [hunk 1]
```diff
@@ -18,86 +18,82 @@
           <entry size="1" alignment="1" />
           <entry size="2" alignment="2" />
           <entry size="4" alignment="4" />
-          <entry size="8" alignment="4" />
+          <entry size="8" alignment="8" />
```
Comment (beardypig) [inline data/languages/r5900.cspec]: Do they definitely have to be 8 byte aligned?

File: build.gradle [hunk 2]
```diff
@@ -31,3 +31,13 @@ else {
 	throw new GradleException("GHIDRA_INSTALL_DIR is not defined!")
 }
 //----------------------END "DO NOT MODIFY" SECTION-------------------------------
+
+def sleigh = new File(ghidraInstallDir).getCanonicalPath() + "/support/sleigh.bat"
```
Comment (beardypig) [inline build.gradle]: Unfortunately this won't work on an OS other than Windows... To compile the sleigh files you use `ant` in the `data` directory (which build.xml is).

File: build.gradle [hunk 3]
```diff
@@ -31,3 +31,21 @@ else {
...
+
+def sleigh_path
+if (OperatingSystem.current() == OperatingSystem.WINDOWS)
+	sleigh_path = "/support/sleigh.bat"
+else
+	sleigh_path = "/support/sleigh"
+
+def sleigh = new File(ghidraInstallDir).getCanonicalPath() + sleigh_path
+task runSleigh(type:Exec) {
+    commandLine sleigh, "-a", "data/languages"
```
Comment (beardypig) [inline build.gradle]: Is compiling the `sla` files required? Doesn't Ghidra do it when you load the extension?

Comment (astrelsky) [reply build.gradle]: It does now. I don't think ghidra did initially. Until any java code is added build.gradle shouldn't be necessary.

Comment (beardypig) [reply build.gradle]: Ah, perhaps. I will check on an older version. The gradle file is used to package the extension in to a zip too.

**Comment by beardypig** [comment]
Do you think this is good to merge?

File: data/languages/r5900.cspec (hunk 1 above)
Comment (astrelsky) [reply data/languages/r5900.cspec]: 100% positive that longs are 8 bytes and 8 byte aligned.

**Comment by astrelsky** [comment]
> Do you think this is good to merge?

Should be good to go now.

**Comment by astrelsky** [comment]
I realized I misunderstood the bitrange assignments. Don't merge this yet I have to go through and fix it. I thought it was like a slice but it is really register[{start_bit}, {range_length}]
Also they are bits, not bytes. I have work to do.

**Comment by astrelsky** [comment]
Ok, I have rewritten them and simplified them as best as possible. For library functions that were written in assembly with the MMI instructions you may see a lot of "CONCAT" but it is just the nature of what is going on and the varnode size restriction.

lq and sq produce a warning about the inaccurate size, but they decompile correctly otherwise.

**Comment by beardypig** [comment]
@astrelsky this needs a rebase too :(

**Comment by astrelsky** [comment]
I handled the conflicts as well as did some updates. It is still possible to encounter CONCAT at times. When this is encountered it is usually because it was a hand-written asm routine instead of compiler optimized.

**Comment by beardypig** [comment]
Cool. I don't think the `CONCAT`s are much of a problem :)

**Pull request merged**